fe/fe-core/src/main/java/org/apache/doris/catalog/authorizer/RangerHiveAccessController.java

```diff
@@ -0,0 +1,188 @@
+import org.apache.doris.common.AuthorizationException;
+import org.apache.doris.mysql.privilege.CatalogAccessController;
+import org.apache.doris.mysql.privilege.PrivPredicate;
+
+import org.apache.hadoop.hive.ql.security.authorization.plugin.HiveAccessControlException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.ranger.plugin.policyengine.RangerAccessRequest;
+import org.apache.ranger.plugin.policyengine.RangerAccessRequestImpl;
+import org.apache.ranger.plugin.policyengine.RangerAccessResult;
+import org.apache.ranger.plugin.policyengine.RangerPolicyEngine;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class RangerHiveAccessController implements CatalogAccessController {
+    public static final String CLIENT_TYPE_DORIS = "doris";
+    private static final Logger LOG = LogManager.getLogger(RangerHiveAccessController.class);
+    private static volatile RangerHivePlugin hivePlugin = null;
+
+    public RangerHiveAccessController(Map<String, String> properties) {
+        String serviceName = properties.get("ranger.service.name");
+        hivePlugin = new RangerHivePlugin(serviceName);
+    }
+
+    private RangerAccessRequestImpl createRequest(UserIdentity currentUser, HiveAccessType accessType) {
+        RangerAccessRequestImpl request = new RangerAccessRequestImpl();
+        request.setUser(currentUser.getQualifiedUser());
+        request.setUserRoles(currentUser.getRoles());
+        request.setAction(accessType.name());
+        if (accessType == HiveAccessType.USE) {
+            request.setAccessType(RangerPolicyEngine.ANY_ACCESS);
+        } else {
+            request.setAccessType(accessType.name().toLowerCase());
+        }
+        request.setClientIPAddress(currentUser.getHost());
+        request.setClientType(CLIENT_TYPE_DORIS);
+        request.setAccessTime(new Date());
+
+        return request;
+    }
+
+    public void checkPrivileges(UserIdentity currentUser, HiveAccessType accessType,
+                                List<RangerHiveResource> hiveResources) throws AuthorizationException {
+        List<RangerAccessRequest> requests = new ArrayList<>();
+        for (RangerHiveResource resource : hiveResources) {
+            RangerAccessRequestImpl request = createRequest(currentUser, accessType);
+            request.setResource(resource);
+
+            requests.add(request);
+        }
+
+        RangerHiveAuditHandler auditHandler = new RangerHiveAuditHandler(hivePlugin.getConfig());
+        Collection<RangerAccessResult> results = hivePlugin.isAccessAllowed(requests, auditHandler);
+        for (RangerAccessResult result : results) {
+            LOG.debug("match policy:" + result.getPolicyId());
+            LOG.debug("will audit: " + result.toString());
+            auditHandler.processResult(result);
```

Comment:
I saw in `isAccessAllowed()`, it already call the `auditHandler.processResult()`, why do we need to call it again here?

Reply:
fixed